### PR TITLE
[SPARK-24321][SQL] Extract common code from Divide/Remainder to a base trait

### PR DESCRIPTION
**Pull request by rednaxelafx**
## What changes were proposed in this pull request?

Extract common code from `Divide`/`Remainder` to a new base trait, `DivModLike`.

Further refactoring to make `Pmod` work with `DivModLike` is to be done as a separate task.

## How was this patch tested?

Existing tests in `ArithmeticExpressionSuite` covers the functionality.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala [hunk 1]
```diff
@@ -220,30 +220,12 @@ case class Multiply(left: Expression, right: Expression) extends BinaryArithmeti
...
-  override def symbol: String = "/"
-  override def decimalMethod: String = "$div"
   override def nullable: Boolean = true
 
-  private lazy val div: (Any, Any) => Any = dataType match {
-    case ft: FractionalType => ft.fractional.asInstanceOf[Fractional[Any]].div
-  }
-
-  override def eval(input: InternalRow): Any = {
+  protected def evalHelper(input: InternalRow, op: => (Any, Any) => Any): Any = {
```
Comment (rednaxelafx) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala]: Note: `op` is declared as a call-by-name argument to retain the `lazy val` semantics of the original code in `Divide.div`.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala]: nit: not sure which one is better, but in object-oriented world we usually do
```
def eval {
  ....
  evalInternal(input1, input2)
}

protected def evalInternal(input1: Any, input2: Any): Any

```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala [hunk 2]
```diff
@@ -297,13 +279,39 @@ case class Divide(left: Expression, right: Expression) extends BinaryArithmetic
...
+
+  override def symbol: String = "/"
+  override def decimalMethod: String = "$div"
+  override def nullable: Boolean = true
+
+  private lazy val div: (Any, Any) => Any = dataType match {
+    case ft: FractionalType => ft.fractional.asInstanceOf[Fractional[Any]].div
+  }
+
+  override def eval(input: InternalRow): Any = evalHelper(input, div)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala]: following my above comment, here we can write
```
private lazy val div = dataType.asInstanceOf[FractionType].fractional.asInstanceOf[Fractional[Any]].div

override def evalInternal(input1: Any, input2: Any) = div(input1, input2)
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala [hunk 3]
```diff
@@ -313,81 +321,26 @@ case class Divide(left: Expression, right: Expression) extends BinaryArithmetic
...
-      val input1 = left.eval(input)
-      if (input1 == null) {
-        null
-      } else {
-        input1 match {
-          case d: Double => d % input2.asInstanceOf[java.lang.Double]
-          case f: Float => f % input2.asInstanceOf[java.lang.Float]
-          case _ => integral.rem(input1, input2)
-        }
+    evalHelper(input, (input1, input2) => {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala]: and here can be
```
private lazy val mod: (Any, Any) => Any = dataType match {
  case i: IntegralType =>
    val integral = i.integral.asInstanceOf[Integral[Any]]
    (input1, input2) => integral.rem(input1, input2)

  case FloatType =>
    (input1, input2) => input1.asInstanceOf[java.lang.Float] % input2..asInstanceOf[java.lang.Float]

  case DoubleType =>
      (input1, input2) => input1.asInstanceOf[java.lang.Double] % input2..asInstanceOf[java.lang.Double]
}

override def evalInternal(input1: Any, input2: Any) = mod(input1, input2)
```

**Comment by cloud-fan** [comment]
LGTM, also cc @viirya @kiszk

**Comment by kiszk** [comment]
LGTM

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala [hunk 4]
```diff
@@ -297,13 +279,39 @@ case class Divide(left: Expression, right: Expression) extends BinaryArithmetic
...
+       1.0
+  """)
+// scalastyle:on line.size.limit
+case class Divide(left: Expression, right: Expression) extends DivModLike {
+
+  override def inputType: AbstractDataType = TypeCollection(DoubleType, DecimalType)
+
+  override def symbol: String = "/"
+  override def decimalMethod: String = "$div"
+  override def nullable: Boolean = true
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala]: redundant?

**Review by viirya — APPROVED** [review]
LGTM

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala (hunk 4 above)
Comment (rednaxelafx) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala]: Oops! Thanks for catching it!
Addressed in an update.

**Comment by rednaxelafx** [comment]
Updated PR, addressed @cloud-fan and @viirya 's comments.

**Comment by viirya** [comment]
retest this please.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**